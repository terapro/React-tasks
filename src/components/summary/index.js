import React from 'react';

import {Label} from '../common/label/label.js';

import {FilmSort} from '../summary/film-sort/film-sort';

import PropTypes from 'prop-types';

const Summary = (props) => (
  <div className='summary'>
    <div className='summary-wrapper'>
      {props.filmMode? null : <Label content ={props.filmNumber + ' movies found'} />}
      {props.filmMode? null : <FilmSort sortItems={props.sortItems} changeSortItemCallBack = {props.changeSortItemCallBack} />}
      {props.filmMode?  <Label content ={'Films by ' + props.filmModeGenre  + ' genre'} /> : null}

    </div>
  </div>
);

Summary.propTypes = {
  filmModeGenre: PropTypes.string,
  filmMode: PropTypes.bool,
  filmNumber: PropTypes.number,
  sortItems: PropTypes.object,
  changeSortItemCallBack: PropTypes.func
};

export {Summary}