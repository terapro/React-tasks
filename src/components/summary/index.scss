.summary{
  height: 50px;
  width: inherit;
  background: #f3f3f3;
}

.summary-wrapper {
  height: 50px;
  width: $summaryContainerWidth;
  margin: 0 auto;
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-weight: bold;
  user-select: none;
}


