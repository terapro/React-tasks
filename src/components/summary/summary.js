import React from 'react';
import PropTypes from 'prop-types';

import {Label} from 'src/components/common/label/label.js';
import {FilmSort} from 'src/components/summary/film-sort/film-sort';
import {connect} from "react-redux";


const Summary = (props) => {
  const {sortItems, changeSortItemCallBack} = props;
  const {filmMode, searchPhrase, filmNumber, filmModeGenre, searchType} = props;

  let innerContent;

  if (filmMode) {
    innerContent = (
      <div className='summary-wrapper'>
        <Label content ={'Films by ' + filmModeGenre  + ' genre'} />
        <FilmSort sortItems={sortItems} changeSortItemCallBack = {changeSortItemCallBack} />
      </div>
    );
  } else {
    innerContent = (
      <div className='summary-wrapper'>
        <Label content ={'Found '+ filmNumber +' films for \'' + searchPhrase + '\' by ' + searchType} />
        <FilmSort sortItems={sortItems} changeSortItemCallBack = {changeSortItemCallBack} />
      </div>
    );
  }
  return (
    <div className='summary'>
        {innerContent}
    </div>
  );
};

Summary.propTypes = {
  searchPhrase: PropTypes.string,
  filmModeGenre: PropTypes.string,
  filmMode: PropTypes.bool,
  filmNumber: PropTypes.number,
  sortItems: PropTypes.shape({
    list: PropTypes.arrayOf(PropTypes.string),
    active: PropTypes.string
  }).isRequired,
  changeSortItemCallBack: PropTypes.func
};

Summary.defaultProps = {
  searchPhrase: '',
  filmModeGenre: '',
  filmMode: false,
  filmNumber: 0,
  changeSortItemCallBack: () => {}
};


export const ConnectedSummary = connect(
  store =>
    ({
      filmMode: store.mode.film,
      searchPhrase: store.search.phrase,
      filmNumber: store.films.searchList.length,
      filmModeGenre: store.mode.filmModeSettings.genre,
      searchType: store.search.type

    }),
  dispatch =>
    ({


    })

)(Summary);

