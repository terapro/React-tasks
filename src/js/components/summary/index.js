import React from 'react';


const Summary = () => {
    return (<div className="summary">this is a summary</div> );

}

export {Summary}