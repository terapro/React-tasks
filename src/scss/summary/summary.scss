.summary{
  height: 50px;
  width: inherit;
  background: #f3f3f3;
}