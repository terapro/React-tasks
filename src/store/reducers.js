import C from  'src/constants';

export const films = (state={}, action) => {
  switch (action.type) {
    case C.START_SEARCH :
      return {
        ...state,
        searchList: action.payload.searchResult
      };
    case C.OPEN_FILM:
      return {
        ...state,
        currentFilm: action.payload.filmToOpen,
        recommendedList: action.payload.recommendedList,
        currentFilmGenre: action.payload.filmGenre
      };
    case C.CHANGE_SORTING:
      return {
        ...state,
        sortBy: filmsSorting(state.sortBy, action)
      };
    default:
      return state
  }
};
export const mode = (state={}, action) => {
  switch (action.type) {
    case C.OPEN_FILM:
      return {
        ...state,
        search: false,
        film: true
      };
    case C.OPEN_SEARCH:
      return {
        ...state,
        search: true,
        film: false
      };
    default:
      return state
  }
};
export const search = (state={}, action) => {
  switch (action.type) {
    case C.START_SEARCH :

    return {
      phrase: action.payload.searchPhrase,
      type: action.payload.searchType,
    };

    default:
      return state
  }
};

 const filmsSorting = (state={}, action) => {
  switch (action.type) {
    case C.CHANGE_SORTING :
      return {
        ...state,
        chosenParameter: action.payload.chosenParameter
      };
    default:
      return state
  }
};






