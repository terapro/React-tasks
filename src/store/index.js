import { createStore, combineReducers, applyMiddleware } from 'redux';
import {films, mode, search} from 'src/store/reducers';
import stateData from 'src/data/initialState.json'
import C from 'src/constants';
import {resultsToStore, recommendedToStore} from 'src/actions'; // TODO; take another action


/**
 * For API requist to the server
 * @param {string} phrase
 * @param {string} searchBy
 * @param {number} limit
 * @returns {Promise<any>}
 * @constructor
 */
const RequestToServer = (phrase, searchBy, limit=21) =>

  new Promise((resolves, rejects) => {
    console.log(phrase);
    console.log(searchBy);
    const api = `http://react-cdp-api.herokuapp.com/movies?search=${phrase}&searchBy=${searchBy}&limit=${limit}`;
    console.log(api);
    const request = new XMLHttpRequest();
    request.open('GET', api);
    request.onload = () =>
      (request.status === 200) ?
        resolves(JSON.parse(request.response)) :
        reject(Error(request.statusText));
    request.onerror = (err) => rejects(err);
    request.send();
  });


const logger = store => next => action => {

  console.log('Before ', action.type);

  // Starting API request for Search mode
  if(action.type == C.START_SEARCH) {// Request to the server
    RequestToServer(action.payload.searchPhrase, action.payload.searchType).then(
      info => store.dispatch(resultsToStore(info.data)),
      err => console.error(
        new Error("cannot load films from server"))
    );
  }
  // Starting API request for Film mode
  if(action.type == C.OPEN_FILM) {// Request to the server
    RequestToServer(action.payload.filmGenre, action.payload.searchType).then(
      info => store.dispatch(recommendedToStore(info.data)),
      err => console.error(
        new Error("cannot load films from server"))
    );
  }

  const returnVal = next(action);
  console.log('After ', action.type);
  return returnVal;
};


const store = createStore(
  combineReducers({films, mode, search}),
  (localStorage['redux-store'])? JSON.parse(localStorage['redux-store']) : stateData,
  applyMiddleware(logger)
);


store.subscribe(() => {
  localStorage['redux-store'] = JSON.stringify(store.getState())
});

//console.log(store.getState());
localStorage['redux-store'] = JSON.stringify(store.getState());

export default store;