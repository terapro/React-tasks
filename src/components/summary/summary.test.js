import React from 'react';
import store from "src/store";
import {Provider} from 'react-redux';
import {render} from 'enzyme';
import {Summary} from 'src/components/summary/summary';

const mockSortBy1 = {
  list: ['release date', 'rating'],
  active: 'release date'
};
const mockSortBy2 = {
  list: ['release date', 'rating'],
  active: 'release date'
};

const mockSearchPhrase = 'Jack';
const mockFilmNumber = 5;
const mockFilmModeGenre = 'Action';
const mockSearchType = 'title';

const mockFunc = jest.fn();

describe('<TypeSection />', () => {
  describe('Render', () => {
    it('should render the component in the film mode with sorting by release date', () => {
      const wrapper = render(<Provider store={store}><Summary filmMode
                                       searchPhrase={mockSearchPhrase}
                                       filmNumber={mockFilmNumber}
                                       filmModeGenre={mockFilmModeGenre}
                                       searchType={mockSearchType}
                                       sortBy={mockSortBy1}
                                       onChangeItem={mockFunc}
      /></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component in the film mode with sorting by rating', () => {
      const wrapper = render(<Provider store={store}><Summary filmMode
                                       searchPhrase={mockSearchPhrase}
                                       filmNumber={mockFilmNumber}
                                       filmModeGenre={mockFilmModeGenre}
                                       searchType={mockSearchType}
                                       sortBy={mockSortBy2}
                                       onChangeItem={mockFunc}
      /></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component in the search mode with sorting by rating', () => {
      const wrapper = render(<Provider store={store}><Summary filmMode={false}
                                       searchPhrase={mockSearchPhrase}
                                       filmNumber={mockFilmNumber}
                                       filmModeGenre={mockFilmModeGenre}
                                       searchType={mockSearchType}
                                       sortBy={mockSortBy2}
                                       onChangeItem={mockFunc}
      /></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component in the search mode with sorting by release date', () => {
      const wrapper = render(<Provider store={store}><Summary filmMode={false}
                                       searchPhrase={mockSearchPhrase}
                                       filmNumber={mockFilmNumber}
                                       filmModeGenre={mockFilmModeGenre}
                                       searchType={mockSearchType}
                                       sortBy={mockSortBy1}
                                       onChangeItem={mockFunc}
      /></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component with default props', () => {
      const wrapper = render(<Provider store={store}><Summary filmMode={false}
                                       searchPhrase={mockSearchPhrase}
                                       filmNumber={mockFilmNumber}
                                       filmModeGenre={mockFilmModeGenre}
                                       searchType={mockSearchType}
                                       sortBy={mockSortBy1}
                                       onChangeItem={mockFunc}
      /></Provider>);
      expect(wrapper).toMatchSnapshot();
    });
  });
});