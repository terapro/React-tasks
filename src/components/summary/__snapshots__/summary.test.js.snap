// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TypeSection /> Render should render the component in the film mode 1`] = `
<div
  className="summary"
>
  <div
    className="summary-wrapper"
  >
    <Label
      content="Films by action genre"
      labelBold={false}
      labelUC={false}
      labelWhite={false}
    />
  </div>
</div>
`;

exports[`<TypeSection /> Render should render the component in the search mode 1`] = `
<div
  className="summary"
>
  <div
    className="summary-wrapper"
  >
    <Label
      content="For 'star wars' 11 movies found"
      labelBold={false}
      labelUC={false}
      labelWhite={false}
    />
    <FilmSort
      changeSortItemCallBack={[MockFunction]}
      sortItems={
        Object {
          "active": "release date",
          "list": Array [
            "release date",
            "rating",
          ],
        }
      }
    />
  </div>
</div>
`;
